File: src/cpp/miru/engine/camera.cpp
#include "miru/engine/transform.h"
#include "miru/engine/sceneobject.h"
#include "miru/engine/color.h"
#include "miru/engine/scene.h"
#include "miru/engine/scenerender.h"
#include "miru/engine/camera.h"

#include "miru/imaging/image.h"

Camera::Camera() : SceneObject()
{

}

Camera::Camera(Camera &camera) : SceneObject(camera)
{
    this->near = camera.near;
    this->far = camera.far;
    this->fov = camera.fov;
}

void Camera::render(Scene* scene, 
                    uint32_t targetRenderWidth, 
                    uint32_t targetRenderHeight, 
                    const char* targetRenderFile)
{
    scene->renderMethod->setupFrame(scene, *this, targetRenderWidth, targetRenderHeight);

    Image* rendered = new Image(targetRenderWidth, targetRenderHeight);

    for (uint32_t y = 0; y < rendered->getHeight(); ++y)
    {
        for (uint32_t x = 0; x < rendered->getWidth(); ++x)
        {
            Color pixelColor = scene->renderMethod->render(scene, x, y);
            rendered->setPixel(x, y, pixelColor);
        }
    }
    
    rendered->saveAsPNG(targetRenderFile);

    delete(rendered);
}

File: src/python/miru/engine/camera.py
from .transform import Transform


class Camera:
    ORTOGRAPHIC = 1
    PERSPECTIVE = 2

    def __init__(self):
        self.mode = Camera.PERSPECTIVE
        self.fov = 60

        self.near = 0.01
        self.far = 1000

        self.transform = Transform()

    @staticmethod
    def parse(data):

        c = Camera()

        if 'mode' in data:
            if data['mode'] == "perspective":
                c.mode = Camera.PERSPECTIVE
            else:
                c.mode = Camera.ORTOGRAPHIC

        if 'fov' in data:
            c.fov = int(data['fov'])

        if 'near' in data:
            c.near = float(data['near'])

        if 'far' in data:
            c.far = float(data['far'])

        if 'transform' in data:
            c.transform.parse(data['transform'])

        return c
